### PR TITLE
[5.4][android] Move to the NDK's unified sysroot

### PR DESCRIPTION
**Pull request by finagolfin**
- Explanation

Since the NDK removes the platforms/ and sysroot/ directories in the latest NDK 22, switch to the unified sysroot in toolchains/llvm/ and take advantage of a bunch of simplification that's now possible.

This is a cherry-pick of #34491 from trunk, only modifying the `build-script-impl` change slightly because of different source indentation than trunk and removing the single C++ Interop test change (that means that test will fail on this branch when tried for Android, but nobody is really running those tests for Android with the release branch).

It is needed because the latest Android NDK 22 finally deprecated the old organization of the Android cross-compilation sysroot, so the 5.4 releases won't build with NDK 22 and future NDKs without this pull.

- Scope

Only affects building for Android, except for one Driver test change, [which was approved by its author](https://github.com/apple/swift/pull/34491#discussion_r550280647).

- SR issue

None

- Risk

Essentially none, as everything is scoped to Android only, except for the single, simple Driver test change.

- Testing

I've run the tests natively on Android from the Jan. 11 source snapshot of this 5.4 branch [with this pull applied](https://github.com/buttaface/termux-packages/blob/swift54/packages/swift/swift-ndk.patch), no problem.

- Reviewer

 @drodriguez

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by buttaface** [comment]
Windows CI failure appears spurious, probably just needs to be re-run:
```
ERROR: [WS-CLEANUP] Cannot delete workspace:
Unable to delete 'S:\jenkins\workspace\swift-PR-windows'.
Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
```

**Comment by shahmishal** [comment]
@swift-ci test Windows

**Comment by shahmishal** [comment]
> Windows CI failure appears spurious, probably just needs to be re-run:
> 
> ```
> ERROR: [WS-CLEANUP] Cannot delete workspace:
> Unable to delete 'S:\jenkins\workspace\swift-PR-windows'.
> Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
> ```

cc: @compnerd

**Comment by buttaface** [comment]
@drodriguez, since you merged this into trunk, put you down here as reviewer too. I think if you approve, this is ready to merge into this branch too.

**Comment by drodriguez** [comment]
Sorry, I didn't understand I had to say something.

This is OK to be merged. The scope is pretty limited to Android. This will allow the 5.4 branch to be more future-proof by allowing using more newer Android NDK if necessary.

**Comment by tomerd** [comment]
seems okay given this is a cherry-pick of https://github.com/apple/swift/pull/34491

**Comment by tkremenek** [comment]
@compnerd @gottesmm: WDYT?

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 1]
```diff
@@ -277,76 +271,32 @@ macro(configure_sdk_unix name architectures)
...
-        set(SWIFT_SDK_ANDROID_ARCH_${arch}_LIBC_INCLUDE_DIRECTORY "${SWIFT_ANDROID_NDK_PATH}/sysroot/usr/include" CACHE STRING "Path to C library headers")
-        set(SWIFT_SDK_ANDROID_ARCH_${arch}_LIBC_ARCHITECTURE_INCLUDE_DIRECTORY "${SWIFT_ANDROID_NDK_PATH}/sysroot/usr/include" CACHE STRING "Path to C library architecture headers")
-      elseif(NOT "${SWIFT_ANDROID_NATIVE_SYSROOT}" STREQUAL "")
-        set(SWIFT_SDK_ANDROID_ARCH_${arch}_LIBC_INCLUDE_DIRECTORY "${SWIFT_ANDROID_NATIVE_SYSROOT}/usr/include" CACHE STRING "Path to C library headers")
-        set(SWIFT_SDK_ANDROID_ARCH_${arch}_LIBC_ARCHITECTURE_INCLUDE_DIRECTORY "${SWIFT_ANDROID_NATIVE_SYSROOT}/usr/include" CACHE STRING "Path to C library architecture headers")
-      else()
-        message(SEND_ERROR "Couldn't find LIBC_INCLUDE_DIRECTORY for Android")
-      endif()
+      swift_android_sysroot(android_sysroot)
+      set(SWIFT_SDK_ANDROID_ARCH_${arch}_PATH "${android_sysroot}")
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:275]: Using `_SYSROOT` instead of `_PATH` would be less confusing IMO.

**Review by compnerd — APPROVED** [review]
I think its safe enough, and is scoped specifically to android.

File: cmake/modules/SwiftConfigureSDK.cmake (hunk 1 above)
Comment (buttaface) [reply cmake/modules/SwiftConfigureSDK.cmake:275]: Maybe, but this is the variable name chosen by Swift for the sysroot for all OS/ARCH combos, [as can be seen earlier in this file](https://github.com/apple/swift/blob/00117e875081c1d9bbf4591eda5853028edc98fd/cmake/modules/SwiftConfigureSDK.cmake#L59) and used elsewhere in the same generic way. If you'd like, maybe that could be changed in a later pull for all OS's.

**Comment by buttaface** [comment]
Ping, do one of the Swift 5.4 branch managers need to sign off on this? This is ready to go in.

**Review by tkremenek — APPROVED** [review]

**Comment by buttaface** [comment]
Thanks, that's [the largest patch I apply to the 5.4 branch when building it for Android](https://github.com/termux/termux-packages/tree/master/packages/swift), ten more to get into trunk (some of those are already in and others are specific to that app).

**Pull request merged**